plugins/optimization-detective/class-od-url-metrics-post-type.php

```diff
@@ -0,0 +1,329 @@
+			return new WP_Error( 'invalid_url_metric', $e->getMessage() );
+		}
+
+		$post_data['post_content'] = wp_json_encode(
+			array_map(
+				static function ( OD_URL_Metric $url_metric ): array {
+					return $url_metric->jsonSerialize();
+				},
+				$group_collection->get_flattened_url_metrics()
+			),
+			JSON_UNESCAPED_SLASHES // No need for escaped slashes since not printed to frontend.
+		);
+
+		$has_kses = false !== has_filter( 'content_save_pre', 'wp_filter_post_kses' );
+		if ( $has_kses ) {
+			// Prevent KSES from corrupting JSON in post_content.
+			kses_remove_filters();
+		}
+
+		$post_data['post_type']   = self::SLUG;
+		$post_data['post_status'] = 'publish';
+		$slashed_post_data        = wp_slash( $post_data );
+		if ( isset( $post_data['ID'] ) ) {
+			$result = wp_update_post( $slashed_post_data, true );
+		} else {
+			$result = wp_insert_post( $slashed_post_data, true );
+		}
+
+		if ( $has_kses ) {
+			kses_init_filters();
+		}
+
+		return $result;
+	}
+
+	/**
+	 * Schedules garbage collection of stale URL Metrics.
+	 *
+	 * @since 0.1.0
+	 */
+	public static function schedule_garbage_collection() {
+		if ( ! is_user_logged_in() ) {
+			return;
+		}
+
+		// Unschedule any existing event which had a differing recurrence.
+		$scheduled_event = wp_get_scheduled_event( self::GC_CRON_EVENT_NAME );
+		if ( $scheduled_event && self::GC_CRON_RECURRENCE !== $scheduled_event->schedule ) {
+			wp_unschedule_event( $scheduled_event->timestamp, self::GC_CRON_EVENT_NAME );
+			$scheduled_event = null;
+		}
+
+		if ( ! $scheduled_event ) {
+			wp_schedule_event( time(), self::GC_CRON_RECURRENCE, self::GC_CRON_EVENT_NAME );
+		}
+	}
+
+	/**
+	 * Deletes posts that have not been modified in the past month.
+	 *
+	 * @since 0.1.0
+	 */
+	public static function delete_stale_posts() {
+		$one_month_ago = gmdate( 'Y-m-d H:i:s', strtotime( '-1 month' ) );
+
+		$query = new WP_Query(
+			array(
+				'post_type'      => self::SLUG,
+				'posts_per_page' => 100,
+				'date_query'     => array(
+					'column' => 'post_modified_gmt',
+					'before' => $one_month_ago,
+				),
+			)
+		);
+
+		foreach ( $query->posts as $post ) {
+			if ( self::SLUG === $post->post_type ) { // Sanity check.
+				wp_delete_post( $post->ID, true );
+			}
+		}
+	}
+
+	/**
+	 * Deletes all URL Metrics posts.
+	 *
+	 * This is used during uninstallation.
+	 *
+	 * @since 0.1.0
+	 */
+	public static function delete_all_posts() {
+		global $wpdb;
+
+		// phpcs:disable WordPress.DB.DirectDatabaseQuery.DirectQuery, WordPress.DB.DirectDatabaseQuery.NoCaching
+
+		// Delete all related post meta for URL Metrics posts.
+		$wpdb->query(
+			$wpdb->prepare(
+				"
+				DELETE meta
+				FROM $wpdb->postmeta AS meta
+					INNER JOIN $wpdb->posts AS posts
+						ON posts.ID = meta.post_id
+				WHERE posts.post_type = %s;
+				",
+				self::SLUG
+			)
+		);
+
+		// Delete all URL Metrics posts.
+		$wpdb->delete(
+			$wpdb->posts,
+			array(
+				'post_type' => self::SLUG,
+			)
+		);
```

Comment:
Since the only plugin code that would be quering these posts is being uninstalled, I didn't think it was worthwhile clearing the cache which could be a significant performance hit.

What last changed value are you referring to?

Reply:
I didn't mean that we should clear (flush) the entire cache, but rather invalidate relevant caches affected by the deletion of these posts.

We need to update the "last changed" value that applies to all WordPress posts, so that `WP_Query` won't serve any stale data. This value is changed whenever any post, of any post type, is changed.

This is trivial to do by calling `wp_cache_set_posts_last_changed()`

Reply:
Added in a620bd0a